File: scripts/network.py
import asyncio
import click
import re
from rich.console import Console
from playwright.async_api import async_playwright, Request, Response
from browser_utils import get_browser_and_page

console = Console()

@click.command()
@click.option(
    "--port",
    type=int,
    default=9222,
    help="Remote debugging port (default: 9222)"
)
@click.option(
    "--type",
    "resource_type",
    type=click.Choice(["all", "xhr", "fetch", "document", "script", "stylesheet", "image", "font", "media"]),
    default="all",
    help="Filter by resource type (default: all)"
)
@click.option(
    "--show-headers",
    is_flag=True,
    help="Show request and response headers"
)
@click.option(
    "--show-body",
    is_flag=True,
    help="Show request and response bodies (only for fetch/xhr)"
)
@click.option(
    "--filter",
    "url_filter",
    type=str,
    default=None,
    help="Filter URLs by regex pattern"
)
def main(port, resource_type, show_headers, show_body, url_filter):
    """Capture network requests from an existing Chrome instance."""
    asyncio.run(capture_network(port, resource_type, show_headers, show_body, url_filter))

async def capture_network(port, resource_type, show_headers, show_body, url_filter):
    async with async_playwright():
        try:
            browser, page = await get_browser_and_page(port)
            if not browser or not page:
                return

            url_pattern = re.compile(url_filter) if url_filter else None
            request_count = 0

            def handle_request(request: Request):
                nonlocal request_count

                if resource_type != "all" and request.resource_type != resource_type:
                    return

                if url_pattern and not url_pattern.search(request.url):
                    return

                request_count += 1

                console.print(f"\n[bold cyan]→ REQUEST #{request_count}[/bold cyan]")
                console.print(f"  [cyan]{request.method:6}[/cyan] [{request.resource_type:10}] {request.url}")

                if show_headers:
                    console.print("  [yellow]Request Headers:[/yellow]")
                    for key, value in request.headers.items():
                        console.print(f"    {key}: {value}")

                if show_body and request.resource_type in ["fetch", "xhr"] and request.post_data:
                    console.print("  [yellow]Request Body:[/yellow]")
                    console.print(f"    {request.post_data}")

            async def handle_response(response: Response):
                if resource_type != "all" and response.request.resource_type != resource_type:
                    return

                if url_pattern and not url_pattern.search(response.url):
                    return

                status_style = "green" if 200 <= response.status < 300 else "red"
                console.print(f"[bold {status_style}]← RESPONSE[/bold {status_style}]")
                console.print(f"  [{status_style}]{response.status}[/{status_style}] {response.url}")

                if show_headers:
                    headers = await response.all_headers()
                    console.print("  [green]Response Headers:[/green]")
                    for key, value in headers.items():
                        console.print(f"    {key}: {value}")

                if show_body and response.request.resource_type in ["fetch", "xhr"]:
                    console.print("  [green]Response Body:[/green]")
                    try:
                        body = await response.text()
                        console.print(f"    {body}")
                    except Exception:
                        try:
                            body_bytes = await response.body()
                            console.print(f"    <binary data, {len(body_bytes)} bytes>")
                        except Exception:
                            console.print("    <failed to read>")

            page.on("request", handle_request)
            page.on("response", handle_response)

            console.print("[bold cyan]Listening for network requests (press Ctrl+C to stop)...[/bold cyan]")

            try:
                while True:
                    await asyncio.sleep(1)
            except KeyboardInterrupt:
                console.print(f"\n\n[bold cyan]Stopped. Total requests: {request_count}[/bold cyan]")

        except Exception as e:
            click.echo(f"Failed to connect or capture: {e}", err=True)

if __name__ == "__main__":
    main()

File: cmd/network.go
package cmd

import (
	"context"
	"fmt"
	"os"
	"os/signal"
	"regexp"
	"strings"
	"sync"
	"syscall"

	"browser-tools/browser"
	"github.com/chromedp/cdproto/network"
	"github.com/chromedp/chromedp"
	flag "github.com/spf13/pflag"
)

const (
	ansiReset  = "\033[0m"
	ansiBold   = "\033[1m"
	ansiDim    = "\033[2m"
	ansiCyan   = "\033[36m"
	ansiGreen  = "\033[32m"
	ansiYellow = "\033[33m"
	ansiRed    = "\033[31m"
	ansiBlue   = "\033[34m"
)

type capturedRequest struct {
	num     int
	id      network.RequestID
	url     string
	method  string
	resType network.ResourceType
	headers network.Headers
	body    string
}

func Network(ctx context.Context, variant string, port int, args []string) {
	fs := flag.NewFlagSet("network", flag.ExitOnError)
	resTypeFlag := fs.String("type", "all", "filter by resource type: all, xhr, fetch, document, script, stylesheet, image, font, media, websocket, other")
	showHeaders := fs.Bool("show-headers", false, "show request and response headers")
	showBody := fs.Bool("show-body", false, "show response body (xhr/fetch only)")
	urlFilter := fs.String("filter", "", "filter URLs by regex")
	fs.Usage = func() {
		fmt.Fprintln(os.Stderr, "Usage: browser-tools network [options]")
		fmt.Fprintln(os.Stderr)
		fs.PrintDefaults()
	}
	if err := fs.Parse(args); err != nil {
		os.Exit(1)
	}

	var urlRe *regexp.Regexp
	if *urlFilter != "" {
		var err error
		urlRe, err = regexp.Compile(*urlFilter)
		if err != nil {
			fmt.Fprintf(os.Stderr, "invalid --filter regex: %v\n", err)
			os.Exit(1)
		}
	}

	typeFilter := strings.ToLower(*resTypeFlag)

	if err := browser.EnsureRunning(variant, port); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	allocCtx, _ := browser.Connect(context.Background(), port)
	tabCtx, _ := browser.ExistingOrNewTab(allocCtx)

	var (
		mu       sync.Mutex
		requests = make(map[network.RequestID]*capturedRequest)
		count    int
	)

	matches := func(url string, rt network.ResourceType) bool {
		if typeFilter != "all" && !strings.EqualFold(string(rt), typeFilter) {
			return false
		}
		if urlRe != nil && !urlRe.MatchString(url) {
			return false
		}
		return true
	}

	statusColor := func(status int64) string {
		switch {
		case status >= 500:
			return ansiRed
		case status >= 400:
			return ansiRed
		case status >= 300:
			return ansiYellow
		default:
			return ansiGreen
		}
	}

	printHeaders := func(label, color string, headers network.Headers) {
		fmt.Printf("  %s%s%s\n", color, label, ansiReset)
		for k, v := range headers {
			fmt.Printf("    %s%s%s: %v\n", ansiDim, k, ansiReset, v)
		}
	}

	chromedp.ListenTarget(tabCtx, func(ev any) {
		switch e := ev.(type) {
		case *network.EventRequestWillBeSent:
			if e.RedirectResponse != nil {
				return
			}
			if !matches(e.Request.URL, e.Type) {
				return
			}

			mu.Lock()
			count++
			n := count
			req := &capturedRequest{
				num:     n,
				id:      e.RequestID,
				url:     e.Request.URL,
				method:  e.Request.Method,
				resType: e.Type,
				headers: e.Request.Headers,
				body:    "",
			}
			if e.Request.HasPostData {
				var parts []string
				for _, entry := range e.Request.PostDataEntries {
					if entry.Bytes != "" {
						parts = append(parts, entry.Bytes)
					}
				}
				req.body = strings.Join(parts, "")
			}
			requests[e.RequestID] = req
			mu.Unlock()

			rt := strings.ToLower(string(e.Type))
			fmt.Printf("\n%s%s→ REQ  #%-4d%s %s%-7s%s %s[%-11s]%s %s\n",
				ansiBold, ansiCyan, n, ansiReset,
				ansiBold, e.Request.Method, ansiReset,
				ansiDim, rt, ansiReset,
				e.Request.URL)

			if *showHeaders {
				printHeaders("Request Headers:", ansiBlue, e.Request.Headers)
			}
			if *showBody && req.body != "" {
				fmt.Printf("  %sRequest Body:%s\n    %s\n", ansiBlue, ansiReset, req.body)
			}

		case *network.EventResponseReceived:
			mu.Lock()
			req, ok := requests[e.RequestID]
			mu.Unlock()
			if !ok {
				return
			}

			sc := statusColor(e.Response.Status)
			fmt.Printf("%s%s← RES  #%-4d%s %s%d %s%s %s\n",
				ansiBold, sc, req.num, ansiReset,
				sc, e.Response.Status, e.Response.StatusText, ansiReset,
				e.Response.URL)

			if *showHeaders {
				printHeaders("Response Headers:", ansiGreen, e.Response.Headers)
			}

		case *network.EventLoadingFinished:
			if !*showBody {
				return
			}
			mu.Lock()
			req, ok := requests[e.RequestID]
			mu.Unlock()
			if !ok {
				return
			}
			rt := strings.ToLower(string(req.resType))
			if rt != "xhr" && rt != "fetch" {
				return
			}

			reqID := e.RequestID
			num := req.num
			go func() {
				var body []byte
				if err := chromedp.Run(tabCtx, chromedp.ActionFunc(func(ctx context.Context) error {
					var err error
					body, err = network.GetResponseBody(reqID).Do(ctx)
					return err
				})); err != nil {
					fmt.Printf("  %s[BODY  #%-4d]%s <error: %v>\n", ansiDim, num, ansiReset, err)
					return
				}
				bodyStr := strings.TrimSpace(string(body))
				if bodyStr == "" {
					return
				}
				fmt.Printf("  %s[BODY  #%-4d]%s %s\n", ansiDim, num, ansiReset, bodyStr)
			}()

		case *network.EventLoadingFailed:
			mu.Lock()
			req, ok := requests[e.RequestID]
			mu.Unlock()
			if !ok {
				return
			}
			fmt.Printf("%s%s✗ FAIL #%-4d%s %s — %s\n",
				ansiBold, ansiRed, req.num, ansiReset,
				req.url, e.ErrorText)
		}
	})

	if err := chromedp.Run(tabCtx, network.Enable()); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "%sCatching network requests (Ctrl+C to stop)...%s\n", ansiDim, ansiReset)

	sig := make(chan os.Signal, 1)
	signal.Notify(sig, os.Interrupt, syscall.SIGTERM)
	<-sig

	mu.Lock()
	total := count
	mu.Unlock()
	fmt.Fprintf(os.Stderr, "\n%sStopped. Total requests captured: %d%s\n", ansiBold, total, ansiReset)
}
